### ISSUE TITLE
How to save and restore the model?

### ISSUE THREAD

**Issue by topgetgithub** [chewxy/gorgonia#170]
How to save and restore the model? And how to use the result model to get the expected output?

**Comment by chewxy** [comment]
You typically want to save the weights of the models, not the model itself. The model itself is typically Go code.

**Comment by topgetgithub** [comment]
I agree. And would you kindly give the equivalent of the code(tensorflow, python format) in Gorgonia below?

x_data = np.array([[1.0, 1.0, 1.0], [1.0, 0.0, 1.0], [1.0, 2.0, 3.0]], dtype=np.float32)
x = tf.placeholder(tf.float32, name="x") 

y_data = np.array([2.0, 1.0, 3.0], dtype=np.float32)
y_ = tf.placeholder(tf.float32, name="y_") 

w1 = tf.Variable(tf.random_normal([rowSize - 3], stddev=0.2), dtype=tf.float32, name="w1")
b1 = tf.Variable(1, dtype=tf.float32, name="b1")

w2 = tf.Variable(tf.random_normal([rowSize - 3], stddev=0.3), dtype=tf.float32, name="w2")
b2 = tf.Variable(0, dtype=tf.float32, name="b2")

y1 = tf.nn.sigmoid((x * w1) + b1)

y2 = tf.reduce_sum(tf.nn.tanh(y1 * w2)) + b2

loss = tf.reduce_sum(tf.square(tf.abs(y2 - y_)))

optimizer = tf.train.RMSPropOptimizer(0.1)

train = optimizer.minimize(loss)

**Comment by chewxy** [comment]
```
func Foo(xData, yData tensor.Tensor) {
g := G.NewGraph()
x := G.NodeFromAny(g, xData, G.WithName("x"))
y := G.NodeFromAny(g, yData, G.WithName("y"))

w1 := G.NewVector(G.WithShape(3), G.Of(G.Float32), G.WithName("w1"), G.WithInit(G.Uniform(1.0)))
b1 := G.NewScalar(G.WithName("b1", G.Of(G.Float32))

w2 := G.NewVector(G.WithShape(3), G.Of(G.Float32), G.WithName("w2"), G.WithInit(G.Uniform(1.0)))
b2 := G.NewScalar(G.WithName("b2", G.Of(G.Float32))

y1 := G.Must(G.Sigmoid(G.Must(G.Add(G.Must(G.Mul(x, w1)), b1))))
y2 := G.Must(G.Sum(G.Must(G.Add(G.Must(G.Tanh(G.Must(G.Mul(y1, w2))), b2))))

loss := G.Must(G.Sum(G.Must(G.Square(G.Must(G.Abs(G.Must(G.Sub(y2, y))))))))

machine := G.NewTapeMachine(g)
optimizer := G.NewRMSPropSolver()
if err := machine.RunAll(); err != nil {
panic(err)
}
if err := optimizer.Step(Nodes{w1, b1, w2, b2}); err != nil{
panic(err)
}
```

**Comment by chewxy** [comment]
I highly discourage the use of `Must()`  though.

**Comment by topgetgithub** [comment]
really appreciate your quickly response. Could you give me an example to show how to feed the value(the parameters of the function)? I have data in the form as:

	inputStr := "218|4.45|19|62.84|11.47|9.47|1.74|5|4.35|0|0.00|5|2.29|3|2.61|3|1.38|63|150|42|58|4|4|1|1|0|0|1|68|31.19|160|73.39|214|98.17|217|99.54|218|100.00|13|5.96|31|14.22|42|19.27|43|19.72|43|19.72|105|16|0|33|8|20|0|0|17|11|117|4|3|0|0|7|22|22|15|3|9|9|0|0|0|0|0|1|16|6|0|0|3|0|0|3|0|0|0|0|0|0|0|0|1|0|0|0|1|0|0|0|0|0|0"

	inputList := strings.Split(inputStr, "|")

	var inputArray []float32 = make([]float32, len(inputList))

	for i, v := range inputList {
		inputArray[i] = StringToFloat32WithDefaultValue(v, 0)
	}

	fmt.Println("input: %v", inputArray)

	fakeInput, _ := tf.NewTensor(inputArray)

**Comment by chewxy** [comment]
What's the shape gonna be?

**Comment by chewxy** [comment]
I'm thinking something like `tensor.New(tensor.WithBacking(inputArray))` would fit your bill quite well

Optionally with shape:
`tensor.New(tensor.WithBacking(inputArray), tensor.WithShape(...))`

**Comment by topgetgithub** [comment]
1, 103
eg. [1, 2, 1, ..., n] 103 items, 1D

**Comment by topgetgithub** [comment]
Ok, I'll try both. Thanks a lot.

**Comment by chewxy** [comment]
I'll close this for now. When you get a new problem feel free to open a new ticket
